wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -223,6 +234,152 @@ newUnsignedTransaction ActiveWallet{..} options accountId payees = runExceptT $
+  where
+    -- Take addresses that correspond to inputs of transaction.
+    extractSrcAddressesFrom :: UnsignedTx -> NonEmpty Address
+    extractSrcAddressesFrom unsignedTx =
+        NonEmpty.map (\(_, outAux) -> txOutAddress . toaOut $ outAux) $ utxOwnedInputs unsignedTx
+
+    buildDerivationPathsFor :: NonEmpty HD.HdAddress
+                            -> HD.HdAccountId
+                            -> NonEmpty [DerivationIndex]
+    buildDerivationPathsFor srcAddresses srcAccountId =
+        flip NonEmpty.map srcAddresses $ \(HD.HdAddress srcAddressId _) ->
+            let HD.HdAccountId _ (HD.HdAccountIx srcAccIndex)  = srcAccountId
+                HD.HdAddressId _ (HD.HdAddressIx srcAddrIndex) = srcAddressId
+            in [srcAccIndex, srcAddrIndex]
+
+    -- Convert 'UnsignedTx' to 'Tx', this regular 'Tx'
+    -- will be signed later (technically - the hash of the
+    -- transaction will be signed).
+    toRegularTx :: UnsignedTx -> Tx
+    toRegularTx unsignedTx = UnsafeTx inputs outputs attribs
+      where
+        inputs  = NonEmpty.map fst $ utxOwnedInputs unsignedTx
+        outputs = NonEmpty.map toaOut $ utxOutputs unsignedTx
+        attribs = utxAttributes unsignedTx
+
+-- | Converts source address to 'HdAddress'.
+-- Since 'srcAddress' is returned by 'newUnsignedTransaction'
+-- we assume that this address is valid so we definitely have
+-- corresponding 'HdAddress'.
+srcAddressToHdAddress :: DB
+                      -> Address
+                      -> HD.HdAddress
+srcAddressToHdAddress db srcAddress =
+    either (error "Impossible happened: source address doesn't have corresponding HD-address!")
+           id
+           (lookupCardanoAddress db srcAddress)
+
+-- | Submits extrenally-signed transaction to the blockchain.
+-- The result of this function is equal to the result of 'pay' function.
+submitSignedTx :: ActiveWallet
+               -> Tx
+               -> NonEmpty (Address, Signature TxSigData, PublicKey)
+               -> IO (Either PaymentError (Tx, TxMeta))
+submitSignedTx aw@ActiveWallet{..} tx srcAddrsWithProofs = do
+    retrying retryPolicy shouldRetry $ \rs -> do
+        -- STEP 1: create witnesses.
+        -- Since we already received inputs signatures with corresponding derived PKs,
+        -- just form witnesses from them.
+        let witnesses = V.fromList . NonEmpty.toList $ flip NonEmpty.map srcAddrsWithProofs $
+                \(_srcAddr, txSig, derivedPK) -> PkWitness derivedPK txSig
+
+        -- STEP 2: make 'TxAux'.
+        let txAux = TxAux tx witnesses
+
+        -- STEP 3: Compute metadata
+        let txId = hash tx
+        -- Currently it's assumed that all source addresses belong to 'the same/ account,
+        -- so we can just take the first source address to find our 'HdAccountId'.
+        let (firstSrcAddress, _, _) = NonEmpty.head srcAddrsWithProofs
+        snapshot <- liftIO $ getWalletSnapshot walletPassive
```

Comment:
It's a minor thing, but as a general good pattern, you should grab the DB snapshot _as soon as you can_, so that it's guaranteed that you see a consistent snapshot of your data for the whole function.